src/internal.c

```diff
@@ -222,6 +225,178 @@ static int SSL_hmac(WOLFSSL* ssl, byte* digest, const byte* in, word32 sz,
+
+        /* pass the log to the client callback*/
+        logCb(ssl, (char*)log);
+
+        XFREE(log, ssl->heap, DYNAMIC_TYPE_SECRET);
+        return 0;
+    }
+#if defined(WOLFSSL_TLS13)
+    static int SessionSecret_callback_Tls13(WOLFSSL* ssl, int id,
+        const unsigned char* secret, int secretSz, void* ctx)
+    {
+        wolfSSL_CTX_keylog_cb_func logCb = NULL;
+        char  label[50];
+        int   labelSz = 0;
+        int   buffSz  = 0;
+        byte* log     = NULL;
+        word32 outSz;
+
+        (void)ctx;
+
+        if (ssl == NULL || secret == NULL || secretSz == 0)
+            return BAD_FUNC_ARG;
+        if (ssl->arrays == NULL || ssl->arrays->clientRandom == NULL)
+            return BAD_FUNC_ARG;
+
+        /* get the user-callback func from CTX*/
+        logCb = ssl->ctx->keyLogCb;
+        if (logCb == NULL)
+            return 0;
+
+        switch (id) {
+            case CLIENT_EARLY_TRAFFIC_SECRET:
+
+                labelSz = sizeof("CLIENT_EARLY_TRAFFIC_SECRET");
+                XSTRNCPY(label,"CLIENT_EARLY_TRAFFIC_SECRET", labelSz);
+                break;
+
+            case CLIENT_HANDSHAKE_TRAFFIC_SECRET:
+
+                labelSz = sizeof("CLIENT_HANDSHAKE_TRAFFIC_SECRET");
+                XSTRNCPY(label, "CLIENT_HANDSHAKE_TRAFFIC_SECRET", labelSz);
+                break;
+
+            case SERVER_HANDSHAKE_TRAFFIC_SECRET:
+
+                labelSz = sizeof("SERVER_HANDSHAKE_TRAFFIC_SECRET");
+                XSTRNCPY(label, "SERVER_HANDSHAKE_TRAFFIC_SECRET", labelSz);
+                break;
+
+            case CLIENT_TRAFFIC_SECRET:
+
+                labelSz = sizeof("CLIENT_TRAFFIC_SECRET_0");
+                XSTRNCPY(label, "CLIENT_TRAFFIC_SECRET_0", labelSz);
+                break;
+
+            case SERVER_TRAFFIC_SECRET:
+
+                labelSz = sizeof("SERVER_TRAFFIC_SECRET_0");
+                XSTRNCPY(label, "SERVER_TRAFFIC_SECRET_0", labelSz);
+                break;
+
+            case EARLY_EXPORTER_SECRET:
+
+                labelSz = sizeof("EARLY_EXPORTER_SECRET");
+                XSTRNCPY(label, "EARLY_EXPORTER_SECRET", labelSz);
+                break;
+
+            case EXPORTER_SECRET:
+
+                labelSz = sizeof("EXPORTER_SECRET");
+                XSTRNCPY(label, "EXPORTER_SECRET", labelSz);
+                break;
+
+            default:
+                return BAD_FUNC_ARG;
+        }
+        /* prepare a log string for passing user callback */
+        buffSz = labelSz + (RAN_LEN * 2) + 1 + secretSz * 2 + 1;
+        log    = XMALLOC(buffSz, ssl->heap, DYNAMIC_TYPE_SECRET);
+        if (log == NULL)
+            return MEMORY_E;
+
+        XMEMSET(log, 0, buffSz);
+        XMEMCPY(log, label, labelSz - 1);     /* put label w/o terminator */
+        XMEMSET(log + labelSz - 1, ' ', 1);               /* '\0' -> ' ' */
+
+        outSz = buffSz - labelSz;
+        Base16_Encode(ssl->arrays->clientRandom, RAN_LEN,
```

Comment:
Need to check return code of Base16_Encode()

Reply:
Added return code check.